File: python/src/highest.py
#!/usr/bin/env python3
import sys
import json
import heapq
import time
import uuid
import psutil
import gc
import threading
from typing import List, Dict, Tuple
from opentelemetry import metrics
from opentelemetry.sdk.metrics import MeterProvider
from opentelemetry.sdk.metrics.export import PeriodicExportingMetricReader
from opentelemetry.exporter.otlp.proto.grpc.metric_exporter import OTLPMetricExporter
from opentelemetry.sdk.resources import Resource

# Generate a unique run ID
run_id = str(uuid.uuid4())[:8]

# Initialize OpenTelemetry metrics
resource = Resource.create({
    "service.name": "highest-scores-python",
    "service.instance.id": "python-impl",
    "run.id": run_id
})

reader = PeriodicExportingMetricReader(
    OTLPMetricExporter(endpoint="otel-collector:4317", insecure=True)
)
provider = MeterProvider(resource=resource, metric_readers=[reader])
metrics.set_meter_provider(provider)
meter = metrics.get_meter("highest-scores")

# Create metrics
processed_lines = meter.create_counter(
    "processed_lines_total",
    description="Number of lines processed",
    unit="1"
)

heap_operations = meter.create_counter(
    "heap_operations_total",
    description="Number of heap operations",
    unit="1"
)

processing_time = meter.create_histogram(
    "processing_time_seconds",
    description="Time taken to process chunks",
    unit="s"
)

# New metrics to match Go implementation
invalid_records = meter.create_counter(
    "invalid_records_total",
    description="Number of invalid records encountered",
    unit="1"
)

bytes_read = meter.create_counter(
    "bytes_read_total",
    description="Number of bytes read from input",
    unit="bytes"
)

gc_stats = meter.create_observable_gauge(
    "gc_stats",
    description="Garbage collection statistics",
    unit="1"
)

cpu_usage = meter.create_observable_gauge(
    "cpu_usage_percent",
    description="CPU usage percentage",
    unit="percent"
)

memory_usage = meter.create_observable_gauge(
    "memory_usage_bytes",
    description="Current memory usage",
    unit="bytes"
)

thread_count = meter.create_observable_gauge(
    "thread_count",
    description="Number of active threads",
    unit="1"
)

def memory_callback(options):
    """Callback for memory usage metric"""
    attributes = {
        "implementation": "python",
        "run_id": run_id
    }
    yield metrics.Observation(psutil.Process().memory_info().rss, attributes)

def thread_callback(options):
    """Callback for thread count metric"""
    attributes = {
        "implementation": "python",
        "run_id": run_id
    }
    yield metrics.Observation(threading.active_count(), attributes)

def gc_callback(options):
    """Callback for GC statistics"""
    stats = gc.get_stats()
    total_collections = sum(s["collections"] for s in stats)
    total_time = sum(s["collected"] for s in stats)  # approximate time based on objects collected
    total_objects = sum(s["collected"] for s in stats)

    base_attributes = {
        "implementation": "python",
        "run_id": run_id,
    }
    
    # Match Go's gc_stats metrics
    yield metrics.Observation(total_collections, {**base_attributes, "stat_type": "gc_cycles"})
    yield metrics.Observation(total_time, {**base_attributes, "stat_type": "gc_pause_ns"})
    yield metrics.Observation(total_objects, {**base_attributes, "stat_type": "heap_objects"})

def cpu_callback(options):
    """Callback for CPU usage"""
    attributes = {
        "implementation": "python",
        "run_id": run_id
    }
    try:
        cpu_percent = psutil.Process().cpu_percent()
        yield metrics.Observation(cpu_percent, attributes)
    except Exception:
        yield metrics.Observation(0.0, attributes)

# Register callbacks
memory_usage.add_callback(memory_callback)
thread_count.add_callback(thread_callback)
gc_stats.add_callback(gc_callback)
cpu_usage.add_callback(cpu_callback)

def parse_line(line: str) -> Tuple[int, str]:
    """Parse a single line of input, returning (score, id)."""
    try:
        # Skip empty lines
        if not line.strip():
            return None
        
        # Split and parse score
        score_str, record_str = line.strip().split(':', 1)
        score = int(score_str.strip())
        
        # Track bytes read
        bytes_read.add(
            len(line.encode('utf-8')), 
            {"implementation": "python", "run_id": run_id}
        )
        
        # Parse and validate record
        record = json.loads(record_str.strip())
        if 'id' not in record:
            invalid_records.add(
                1, 
                {"implementation": "python", "run_id": run_id, "error_type": "missing_id"}
            )
            raise ValueError("Record missing 'id' field")
            
        return (score, record['id'])
    except (ValueError, json.JSONDecodeError) as e:
        error_type = "json_error" if isinstance(e, json.JSONDecodeError) else "value_error"
        invalid_records.add(
            1, 
            {"implementation": "python", "run_id": run_id, "error_type": error_type}
        )
        return None

def get_highest_scores(filepath: str, n: int) -> List[Dict[str, any]]:
    """Process file and return n highest scores using a min-heap."""
    # Use min-heap to keep track of n highest scores
    heap: List[Tuple[int, str]] = []
    chunk_start_time = time.time()
    chunk_size = 0
    
    try:
        with open(filepath, 'r') as f:
            for line in f:
                attributes = {
                    "implementation": "python",
                    "run_id": run_id
                }
                
                processed_lines.add(1, attributes)
                chunk_size += 1
                
                result = parse_line(line)
                if result is None:
                    continue
                    
                score, record_id = result
                
                if len(heap) < n:
                    # Heap not full, add new item
                    heapq.heappush(heap, (score, record_id))
                    heap_operations.add(1, attributes)
                elif score > heap[0][0]:
                    # Score is higher than smallest in heap
                    heapq.heapreplace(heap, (score, record_id))
                    heap_operations.add(2, attributes)
                
                # Record processing time every 10000 lines
                if chunk_size >= 10000:
                    chunk_time = time.time() - chunk_start_time  # Keep in seconds
                    processing_time.record(chunk_time, attributes)
                    chunk_size = 0
                    chunk_start_time = time.time()
                    
    except FileNotFoundError:
        print(f"Error: File '{filepath}' not found", file=sys.stderr)
        sys.exit(1)
    except Exception as e:
        print(f"Error processing file: {str(e)}", file=sys.stderr)
        sys.exit(2)
    
    # Record final chunk if any
    if chunk_size > 0:
        chunk_time = time.time() - chunk_start_time
        processing_time.record(chunk_time, attributes)
    
    # Convert heap to sorted list of dictionaries
    result = [{"score": score, "id": id_} for score, id_ in heap]
    result.sort(key=lambda x: x["score"], reverse=True)
    return result

def main():
    # Validate command line arguments
    if len(sys.argv) != 3:
        print("Usage: python highest.py <input_file> <n>", file=sys.stderr)
        sys.exit(2)
    
    try:
        n = int(sys.argv[2])
        if n <= 0:
            raise ValueError("N must be positive")
    except ValueError:
        print("Error: Second argument must be a positive integer", file=sys.stderr)
        sys.exit(2)
    
    # Get and output results
    result = get_highest_scores(sys.argv[1], n)
    print(json.dumps(result, indent=4))

if __name__ == "__main__":
    main() 

File: go/src/highest.go
package main

import (
	"bufio"
	"container/heap"
	"encoding/json"
	"flag"
	"fmt"
	"log"
	"math/rand"
	"os"
	"runtime"
	"strings"
	"sync"
	"time"

	"net/http"

	"github.com/prometheus/client_golang/prometheus"
	"github.com/prometheus/client_golang/prometheus/promauto"
	"github.com/prometheus/client_golang/prometheus/promhttp"
)

// Record represents a single data record
type Record struct {
	ID      string          `json:"id"`
	RawData json.RawMessage `json:"-"`
	Score   int
}

// MinHeap implementation for maintaining top N scores
type MinHeap []Record

func (h MinHeap) Len() int            { return len(h) }
func (h MinHeap) Less(i, j int) bool  { return h[i].Score < h[j].Score }
func (h MinHeap) Swap(i, j int)       { h[i], h[j] = h[j], h[i] }
func (h *MinHeap) Push(x interface{}) { *h = append(*h, x.(Record)) }
func (h *MinHeap) Pop() interface{} {
	old := *h
	n := len(old)
	x := old[n-1]
	*h = old[0 : n-1]
	return x
}

// Metrics
var (
	labels = prometheus.Labels{
		"implementation": "go",
	}

	processedLines = promauto.NewCounterVec(
		prometheus.CounterOpts{
			Name: "processed_lines_total",
			Help: "Number of lines processed",
		},
		[]string{"implementation", "run_id"},
	)

	heapOperations = promauto.NewCounterVec(
		prometheus.CounterOpts{
			Name: "heap_operations_total",
			Help: "Number of heap operations performed",
		},
		[]string{"implementation", "run_id"},
	)

	processingTime = promauto.NewHistogramVec(
		prometheus.HistogramOpts{
			Name:    "processing_time_seconds",
			Help:    "Time taken to process chunks",
			Buckets: []float64{.001, .005, .01, .025, .05, .075, .1, .25, .5, .75, 1.0, 2.5, 5.0},
		},
		[]string{"implementation", "run_id"},
	)

	memoryUsage = promauto.NewGaugeVec(
		prometheus.GaugeOpts{
			Name: "memory_usage_bytes",
			Help: "Current memory usage in bytes",
		},
		[]string{"implementation", "run_id"},
	)

	threadCount = promauto.NewGaugeVec(
		prometheus.GaugeOpts{
			Name: "thread_count",
			Help: "Number of active threads/goroutines",
		},
		[]string{"implementation", "run_id"},
	)

	// New metrics
	invalidRecords = promauto.NewCounterVec(
		prometheus.CounterOpts{
			Name: "invalid_records_total",
			Help: "Number of invalid records encountered",
		},
		[]string{"implementation", "run_id", "error_type"},
	)

	bytesRead = promauto.NewCounterVec(
		prometheus.CounterOpts{
			Name: "bytes_read_total",
			Help: "Number of bytes read from input",
		},
		[]string{"implementation", "run_id"},
	)

	gcStats = promauto.NewGaugeVec(
		prometheus.GaugeOpts{
			Name: "gc_stats",
			Help: "Garbage collection statistics",
		},
		[]string{"implementation", "run_id", "stat_type"},
	)

	cpuUsage = promauto.NewGaugeVec(
		prometheus.GaugeOpts{
			Name: "cpu_usage_percent",
			Help: "CPU usage percentage",
		},
		[]string{"implementation", "run_id"},
	)
)

// updateResourceMetrics periodically updates resource usage metrics
func updateResourceMetrics(runID string) {
	labels := prometheus.Labels{
		"implementation": "go",
		"run_id":         runID,
	}

	for {
		var m runtime.MemStats
		runtime.ReadMemStats(&m)

		// Memory metrics
		memoryUsage.With(labels).Set(float64(m.Alloc))
		threadCount.With(labels).Set(float64(runtime.NumGoroutine()))

		// GC stats
		gcLabels := prometheus.Labels{
			"implementation": "go",
			"run_id":         runID,
			"stat_type":      "gc_cycles",
		}
		gcStats.With(gcLabels).Set(float64(m.NumGC))

		gcLabels["stat_type"] = "gc_pause_ns"
		gcStats.With(gcLabels).Set(float64(m.PauseTotalNs))

		gcLabels["stat_type"] = "heap_objects"
		gcStats.With(gcLabels).Set(float64(m.HeapObjects))

		// CPU usage (simplified)
		var cpuUsageValue float64
		if info, err := getCPUUsage(); err == nil {
			cpuUsageValue = info
		}
		cpuUsage.With(labels).Set(cpuUsageValue)

		time.Sleep(time.Second)
	}
}

func getCPUUsage() (float64, error) {
	// This is a simplified version - in production you'd want to use something like gopsutil
	// to get more accurate CPU metrics
	return 0.0, nil
}

// processChunk processes a chunk of the file
func processChunk(chunk []string, n int, runID string) []Record {
	h := &MinHeap{}
	heap.Init(h)
	startTime := time.Now()

	labels := prometheus.Labels{
		"implementation": "go",
		"run_id":         runID,
	}

	for _, line := range chunk {
		processedLines.With(labels).Inc()

		if strings.TrimSpace(line) == "" {
			continue
		}

		parts := strings.SplitN(line, ":", 2)
		if len(parts) != 2 {
			continue
		}

		var score int
		if _, err := fmt.Sscanf(parts[0], "%d", &score); err != nil {
			continue
		}

		var record Record
		if err := json.Unmarshal([]byte(parts[1]), &record); err != nil {
			continue
		}
		record.Score = score

		if h.Len() < n {
			heap.Push(h, record)
			heapOperations.With(labels).Inc()
		} else if score > (*h)[0].Score {
			heap.Pop(h)
			heap.Push(h, record)
			heapOperations.With(labels).Add(2) // Count both pop and push
		}
	}

	// Record processing time for the chunk
	processingTime.With(labels).Observe(time.Since(startTime).Seconds())

	result := make([]Record, h.Len())
	for i := len(result) - 1; i >= 0; i-- {
		result[i] = heap.Pop(h).(Record)
		heapOperations.With(labels).Inc()
	}
	return result
}

func main() {
	// Generate a unique run ID
	rand.Seed(time.Now().UnixNano())
	runID := fmt.Sprintf("%08x", rand.Int31())

	// Start Prometheus metrics server
	go func() {
		http.Handle("/metrics", promhttp.Handler())
		log.Fatal(http.ListenAndServe(":8001", nil))
	}()

	// Start resource metrics collection
	go updateResourceMetrics(runID)

	n := flag.Int("n", 5, "number of top scores to find")
	filename := flag.String("file", "", "input file path")
	flag.Parse()

	if *filename == "" {
		log.Fatal("Please provide an input file")
	}

	file, err := os.Open(*filename)
	if err != nil {
		log.Fatal(err)
	}
	defer file.Close()

	// Determine chunk size and number of workers
	numWorkers := runtime.NumCPU()
	chunkSize := 10000 // Adjust based on testing

	// Create channels for work distribution
	chunks := make(chan []string, numWorkers)
	results := make(chan []Record, numWorkers)
	var wg sync.WaitGroup

	// Start workers
	for i := 0; i < numWorkers; i++ {
		wg.Add(1)
		go func() {
			defer wg.Done()
			for chunk := range chunks {
				results <- processChunk(chunk, *n, runID)
			}
		}()
	}

	// Read and distribute chunks
	go func() {
		scanner := bufio.NewScanner(file)
		currentChunk := make([]string, 0, chunkSize)

		for scanner.Scan() {
			currentChunk = append(currentChunk, scanner.Text())
			if len(currentChunk) == chunkSize {
				chunks <- currentChunk
				currentChunk = make([]string, 0, chunkSize)
			}
		}

		if len(currentChunk) > 0 {
			chunks <- currentChunk
		}
		close(chunks)
	}()

	// Collect and merge results
	go func() {
		wg.Wait()
		close(results)
	}()

	// Merge results using a min-heap
	finalHeap := &MinHeap{}
	heap.Init(finalHeap)

	for result := range results {
		for _, record := range result {
			if finalHeap.Len() < *n {
				heap.Push(finalHeap, record)
			} else if record.Score > (*finalHeap)[0].Score {
				heap.Pop(finalHeap)
				heap.Push(finalHeap, record)
			}
		}
	}

	// Convert to final output format
	type OutputRecord struct {
		Score int    `json:"score"`
		ID    string `json:"id"`
	}

	output := make([]OutputRecord, finalHeap.Len())
	for i := len(output) - 1; i >= 0; i-- {
		record := heap.Pop(finalHeap).(Record)
		output[i] = OutputRecord{Score: record.Score, ID: record.ID}
	}

	// Output JSON
	jsonOutput, err := json.MarshalIndent(output, "", "    ")
	if err != nil {
		log.Fatal(err)
	}
	fmt.Println(string(jsonOutput))
}
